src/main/java/drukier/maze/Maze.java

```diff
@@ -1,170 +1,222 @@
+                    break;
+                case LEFT:
+                    nextX--;
+                    doneDirection++;
+                    break;
+            }
+
+            if ((mazeContains(nextX, nextY)) && (!(getMazeCell(nextX, nextY).isVisited()))) {
 
-        ArrayList<MazeCell> neighbors = null;
+                    nextCell = getMazeCell(nextX, nextY);
+                    current.setNext(nextCell);
+                    removeWalls(current, nextCell, direction);
+                    success = true;
+                }
+                //reset current and get next direction
+                else {
+                    nextX = current.getCellX();
+                    nextY = current.getCellY();
+                    direction = nextDirection();
+                }
 
-        MazeCell top = null;
-        MazeCell bottom = null;
-        MazeCell left = null;
-        MazeCell right = null;
 
 
-        if ((cell.getCellX() - 1) > 0 && (cell.getCellX() - 1) < mazeHeight) {
-            top = new MazeCell(cell.getCellX() - 1, cell.getCellY());
         }
+        if (!current.isVisited()) {
+            visitedCells++;
+            current.setVisited(true);
+        }
+        return success;
+    }
+
+    private boolean mazeContains(int x, int y) {
+        boolean valid = false;
+        //TODO switch?
+        if ((x < mazeWidth) && (y < mazeHeight)) {
+            if ((x >= 0) && (y >= 0)) {
+                valid = true;
+            }
+        }
+        return valid;
+    }
 
-        if ((cell.getCellX() + 1) > 0 && (cell.getCellX() + 1) < mazeHeight) {
-            bottom = new MazeCell(cell.getCellX() + 1, cell.getCellY());
+        private void removeWalls (MazeCell current, MazeCell next, Direction direction){
+            switch (direction) {
+                case UP:
+                    current.setnWall(false);
+                    next.setsWall(false);
+                    break;
+                case DOWN:
+                    current.setsWall(false);
+                    next.setnWall(false);
+                    break;
+                case RIGHT:
+                    current.seteWall(false);
+                    next.setwWall(false);
+                    break;
+                case LEFT:
+                    current.setwWall(false);
+                    next.seteWall(false);
+                    break;
+            }
         }
 
-        if ((cell.getCellY() - 1) > 0 && (cell.getCellY() - 1) < mazeHeight) {
-            left = new MazeCell(cell.getCellX(), cell.getCellY() - 1);
+        private Direction getDirection () {
+
+            return Direction.values()[random.nextInt(Direction.values().length)];
         }
 
-        if ((cell.getCellX() + 1) > 0 && (cell.getCellX() + 1) < mazeHeight) {
-            right = new MazeCell(cell.getCellX(), cell.getCellY() + 1);
+        public Direction nextDirection () {
+            return direction.values()[(direction.ordinal() + 1) % direction.values().length];
```

Comment:
You start off at a random direction, so `nextDirection()` won't cycle through all the directions

Reply:
it does. it circles back